### ISSUE TITLE
read_html ignores paragraphs in table cells

### ISSUE THREAD

**Issue by sasan00** [pandas-dev/pandas#24766]
#### Code Sample, a copy-pastable example if possible

```python
# Your code here
import pandas as pd

html = """
<html>
<body>
<table>
    <tr>
        <td>
            <p>Field 1</p>
            <p>Field 2</p>
        </td>
        <td>
            <p>Value 1</p>
            <p>Value 2</p>
        </td>
    </tr>
</table>
</body>
</html>
"""

tables = pd.read_html(html)
print(tables[0].iat[0, 0])
```
#### Problem description
In the current implementation, the p tags are ignored, and therefore it's not possible to infer that field 1 has value 1 and field 2 has value 2.

#### Expected Output
```python
tables[0].iat[0, 0] == r'Field 1\nField 2'
tables[0].iat[0, 1] == r'Value 1\nValue 2'
```
#### Output of ``pd.show_versions()``
<details>

[paste the output of ``pd.show_versions()`` here below this line]
INSTALLED VERSIONS
------------------
commit: None
python: 3.7.1.final.0
python-bits: 64
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 63 Stepping 2, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

pandas: 0.23.4
pytest: None
pip: 18.1
setuptools: 40.6.3
Cython: None
numpy: 1.15.4
scipy: None
pyarrow: None
xarray: None
IPython: None
sphinx: None
patsy: None
dateutil: 2.7.5
pytz: 2018.9
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: None
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: 4.3.0
bs4: 4.7.1
html5lib: 1.0.1
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: None
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

**Comment by TomAugspurger** [comment]
Can you write the exact expected output?

**Comment by sasan00** [comment]
I have updated the issue with the requested information. Couldn't find a way to remove the "Needs Info" label.

**Comment by TomAugspurger** [comment]
Thanks. Can you check if the HTML parsing libraries (lxml, bs4) typically convert p tags to newlines? Do they provide options to do that normally?

**Comment by sasan00** [comment]
That wouldn't help as the below example shows:
```python
import pandas as pd

html = """
<html>
<body>
<table>
    <tr>
        <td>
            Field 1
            
            Field 2
        </td>
        <td>
            Value 1
            
            Value 2
        </td>
    </tr>
</table>
</body>
</html>
"""

tables = pd.read_html(html)
print(tables[0].iat[0, 0])
```
Still returns "Field 1  Field 2"

**Comment by TomAugspurger** [comment]
I'm just wondering if our behavior matches the expected behavior of the
underlying parsing libraries, and whether they have ways of dealing with
it. Presumably they've had requests for similar features around whitespace
normalization.

**Comment by sasan00** [comment]
lxml respects whitespaces.

```python
import pandas as pd
from lxml.etree import fromstring
from lxml.html import HTMLParser

html = """
<html>
<body>
<table>
    <tr>
        <td>
            Field 1
            
            Field 2
        </td>
        <td>
            Value 1
            
            Value 2
        </td>
    </tr>
</table>
</body>
</html>
"""

tables = pd.read_html(html)
print(tables[0].iat[0, 0])
parser = HTMLParser()
root = fromstring(html, parser)
for elem in root.iter('td'):
    print(repr(elem.text))
```
Result:

Field 1  Field 2
'\n            Field 1\n            \n            Field 2\n        '
'\n            Value 1\n            \n            Value 2\n        '

**Comment by TomAugspurger** [comment]
Thanks. Can you check if pandas explicitly strips / normalizes whitespace
in read_html then? If so, this would be a good parameter to add to
read_html.

**Comment by sasan00** [comment]
Yes. In `_parse_raw_data`, calls are made to `_remove_whitespace `for each column in each row using the default value of the `regex` argument which is `_RE_WHITESPACE` whose value is `re.compile(r'[\r\n]+|\s{2,}')`.

**Comment by TomAugspurger** [comment]
Thanks for investigating. I think an option to disable that behavior makes sense.

You've given two examples now, one with newlines in the text, and one with `<p>` tags. Do you expect to normalize the `<p>` tags to newlines, so that the two would give the same output? Do we have any prior art to copy here?

**Comment by sasan00** [comment]
I think adding an extra argument that provides the function that takes the raw text of a cell, and returned the "cleaned up" version would be best. Its default value would be `_remove_whitespace` to preserve the existing functionality.

**Comment by markmbaum** [comment]
Hi, wondering if this issue was ever resolved? In my case, I have a `<ul>` inside the HTML table and all the elements of each list are squished together after the table is parsed by `read_html`.

**Comment by Derekt2** [comment]
still encountering this bug.

**Comment by jreback** [comment]
@Derekt2 this is open 
you are welcome to submit a pull request to patch

**Comment by SuryaThiru** [comment]
Hi. I'm experiencing the same bug with newlines in cells. I see that there are some existing contributions, will they be merged?

**Comment by RomainL972** [comment]
take
